### PR TITLE
Add support for experiemental hermetic execution mode to TaskRuns

### PR DESCRIPTION
**Pull request by priyawadhwa**
# Changes

This PR adds support for an experimental hermetic execution mode. If users specify this on their TaskRun, then all user step containers are run without network access. Any containers created or injected by tekton are not affected, and sidecar containers are not affected (since they aren't controller by the entrypointer, we can't make them hermetic as of now).

This PR also adds an integration test to make sure that network access isn't available when hermetic mode is enabled!

TEP: https://github.com/tektoncd/community/blob/main/teps/0025-hermekton.md

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

Add support for experimental hermetic execution mode to TaskRuns

```release-note
Add support for experimental hermetic execution mode to TaskRuns
```

cc @dlorenc

### CODE REVIEW

File: cmd/entrypoint/runner.go [hunk 1]
```diff
@@ -58,6 +59,31 @@ func (rr *realRunner) Run(ctx context.Context, args ...string) error {
 	// main process and all children
 	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
 
+	if os.Getenv("TEKTON_RESOURCE_NAME") == "" && os.Getenv(pod.HermeticEnvVar) == "1" {
+		cmd.SysProcAttr.Cloneflags = syscall.CLONE_NEWNS |
```
Comment (dlorenc) [inline cmd/entrypoint/runner.go:63]: I think you can call the "dropNetworking" function instead - I actually did commit that part!

https://github.com/tektoncd/pipeline/blob/1f5980f8c8a05b106687cfa3e5b3193c213cb66e/cmd/entrypoint/namespaces_linux.go#L25

File: pkg/apis/pipeline/v1beta1/taskrun_types.go [hunk 2]
```diff
@@ -66,6 +66,10 @@ type TaskRunSpec struct {
 	// Workspaces is a list of WorkspaceBindings from volumes to workspaces.
 	// +optional
 	Workspaces []WorkspaceBinding `json:"workspaces,omitempty"`
+	// ExecutionMode specifies the execution mode to run user-specified containers in
+	// (tekton injected containers will not run with this policy)
+	// +optional
+	ExecutionMode ExecutionMode `json:"executionMode,omitempty"`
```
Comment (dlorenc) [inline pkg/apis/pipeline/v1beta1/taskrun_types.go:72]: We had discussed including this as an annotation (or maybe only in the v1alpha1 API only)? first, to indicate that it's experimental: https://github.com/tektoncd/community/blob/main/teps/0025-hermekton.md#implementation-plan

I'm not quite sure what the best approach would be here. The API docs don't really say either: https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md
Probably a good topic for the API working group.

File: pkg/pod/pod.go [hunk 3]
```diff
@@ -168,6 +171,15 @@ func (b *Builder) Build(ctx context.Context, taskRun *v1beta1.TaskRun, taskSpec
 		}
 	}
 
+	// Add env var if hermetic execution was requested
+	if taskRun.Spec.ExecutionMode.Hermetic {
+		for i, s := range stepContainers {
+			// Add it at the end so it overrides
+			env := append(s.Env, corev1.EnvVar{Name: HermeticEnvVar, Value: "1"})
```
Comment (dlorenc) [inline pkg/pod/pod.go:178]: The order here is very important - it has to come at the end. A few unit tests would be useful to make sure people can't override it with their own env vars. We might also want to make the variable "TEKTON_HERMETIC" to indicate that it's a reserved word.

File: pkg/apis/pipeline/v1beta1/taskrun_types.go (hunk 2 above)
Comment (afrittoli) [reply pkg/apis/pipeline/v1beta1/taskrun_types.go:72]: We now have an feature gated API fields with related testing helpers that can be used to introduce "alpha" level features into the beta API: https://github.com/tektoncd/pipeline/tree/main/docs/developers#adding-feature-gated-api-fields

**Comment by afrittoli** [comment]
In terms of documentation, it feels to me like hermetic execution mode would deserve a dedicated file in the `docs` folder.
We could then link to it from the various resource docs, `taskrun.md` for this PR.

File: pkg/apis/pipeline/v1beta1/taskrun_types.go (hunk 2 above)
Comment (vdemeester) [reply pkg/apis/pipeline/v1beta1/taskrun_types.go:72]: Agreed on discussing this in the API working group. We should try to not add a new field at first (I tend to like the annotation approach) ; and if we do, we should *hide* it behind the alpha api feature flag (cc @sbwsg )

Comment (dlorenc) [reply pkg/apis/pipeline/v1beta1/taskrun_types.go:72]: +1, so something like:

* annotation behind alpha feature flag
* beta field behind alpha feature flag
* beta field without feature flag ?

File: test/hermetic_taskrun_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+func taskRun(name, namespace string, hermetic bool) *v1beta1.TaskRun {
+	return &v1beta1.TaskRun{
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace},
+		Spec: v1beta1.TaskRunSpec{
+			ExecutionMode: v1beta1.ExecutionMode{
+				Hermetic: hermetic,
+			},
+			Timeout: &metav1.Duration{Duration: time.Minute},
+			TaskSpec: &v1beta1.TaskSpec{
+				Steps: []v1beta1.Step{
```
Comment (sbwsg) [inline test/hermetic_taskrun_test.go:73]: Suggest adding another test to check that `Sidecars` are prevented from accessing the network too.

One slight wrinkle will be that a failing Sidecar doesn't result in a failed TaskRun. Checking the container status might be the way to go there?

File: test/hermetic_taskrun_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	regularTaskRun := taskRun(regularTaskRunName, namespace, false)
+	t.Logf("Creating TaskRun %s, hermetic=false", regularTaskRunName)
+	if _, err := c.TaskRunClient.Create(ctx, regularTaskRun, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create TaskRun `%s`: %s", regularTaskRunName, err)
+	}
+	if err := WaitForTaskRunState(ctx, c, regularTaskRunName, Succeed(regularTaskRunName), "TaskRunCompleted"); err != nil {
+		t.Fatalf("Error waiting for TaskRun %s to finish: %s", regularTaskRunName, err)
+	}
+
+	// now, run the task mode with hermetic mode
```
Comment (sbwsg) [inline test/hermetic_taskrun_test.go:51]: Would be great to migrate these comments into a docstring on the test func explaining what it does.

File: test/README.md [hunk 6]
```diff
@@ -12,7 +12,7 @@ go test ./...
 # Integration tests (against your current kube cluster)
 go test -v -count=1 -tags=e2e -timeout=20m ./test
 
-#conformance tests  (against your current kube cluster)
+# Conformance tests  (against your current kube cluster)
```
Comment (sbwsg) [inline test/README.md:15]: 👍  thank you!

File: cmd/entrypoint/runner.go (hunk 1 above)
Comment (priyawadhwa) [reply cmd/entrypoint/runner.go]: oooh ty!

File: test/hermetic_taskrun_test.go (hunk 5 above)
Comment (priyawadhwa) [reply test/hermetic_taskrun_test.go:54]: done!

File: pkg/pod/pod.go (hunk 3 above)
Comment (priyawadhwa) [reply pkg/pod/pod.go]: SGTM

File: pkg/apis/pipeline/v1beta1/taskrun_types.go (hunk 2 above)
Comment (priyawadhwa) [reply pkg/apis/pipeline/v1beta1/taskrun_types.go]: sgtm, I switched to an annotation and put it behind the alpha feature flag

File: pkg/pod/pod.go [hunk 7]
```diff
@@ -168,6 +171,15 @@ func (b *Builder) Build(ctx context.Context, taskRun *v1beta1.TaskRun, taskSpec
 		}
 	}
 
+	// Add env var if hermetic execution was requested
+	if taskRun.Spec.ExecutionMode.Hermetic {
+		for i, s := range stepContainers {
```
[the remark this answers is not in the document]
Comment (priyawadhwa) [reply pkg/pod/pod.go:181]: I think for now this alpha feature can only cover step containers, since the sidecar containers aren't controlled by the entrypointer (https://github.com/tektoncd/pipeline/issues/1569) -- without the entrypointer I'm not sure how to enforce a hermetic environment on the sidecar containers. 

If anyone has any suggestions pls let me know!

[the remark this answers is not in the document]
Comment (sbwsg) [reply pkg/pod/pod.go:181]: Oh I see!  Good point.

[the remark this answers is not in the document]
Comment (dlorenc) [reply pkg/pod/pod.go:181]: Ugh, I was hoping the sidecar KEP would be started by now but it looks like it missed the 1.22 train as well. A couple suggestions in the meantime:

* do nothing, document that sidecars are not hermetic
* disallow sidecars when hermetic mode is enabled

I can see pros and cons of each, let's start with 1. for alpha and get some feedback?

File: pkg/pod/pod.go [hunk 8]
```diff
@@ -42,6 +42,13 @@ const (
 
 	// TaskRunLabelKey is the name of the label added to the Pod to identify the TaskRun
 	TaskRunLabelKey = pipeline.GroupName + pipeline.TaskRunLabelKey
+
+	// TektonHermeticEnvVar is the env var we set in containers to indicate they should be run hermetically
+	TektonHermeticEnvVar = "HERMETIC"
```
Comment (sbwsg) [inline pkg/pod/pod.go:47]: I might have missed a separate convo elsewhere; do we still want this to be `"TEKTON_HERMETIC"` to indicate it's a reserved name?

**Review by sbwsg — COMMENTED** [review]
One more suggestion but otherwise this lgtm: We have [the "Alpha Features" section in `docs/install.md`](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features) - we could mention Hermetic builds there, and link to `docs/hermetic.md`?

**Comment by sbwsg** [comment]
Awesome, cheers!

/lgtm

**Comment by sbwsg** [comment]
Integration tests should be back in working order once https://github.com/tektoncd/pipeline/pull/3960 is merged.  You'll need to rebase once that's in.

**Comment by priyawadhwa** [comment]
Thanks @sbwsg!

File: test/hermetic_taskrun_test.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TestHermeticTaskRun make sure that the hermetic execution mode actually drops network from a TaskRun step
+// it does this by first running the TaskRun normally to make sure it passes
+// Then, it enables hermetic mode and makes sure the same TaskRun fails because it no longer has access to a network.
+func TestHermeticTaskRun(t *testing.T) {
+	ctx := context.Background()
```
Comment (chmouel) [inline test/hermetic_taskrun_test.go:35]: I think we had that issue before for other test calling dropNetworking() 

if this test is running on a dev laptop environment that is not Linux, the test will fail. 

Since calling to dropNetworking which is only available on Linux https://github.com/tektoncd/pipeline/blob/567dce3cc/cmd/entrypoint/namespaces.go#L10 would panic()

Comment (chmouel) [reply test/hermetic_taskrun_test.go:35]: ah nm this is the e2e tests, i didn't realize so forget my comment :)

Comment (vdemeester) [reply test/hermetic_taskrun_test.go:35]: Right @chmouel this is not gonna be a problem. One problem though might be running this test against OpenShift or a cluster that enforce some lower privileges (default to user, drop some privileges) that might make the calls in `dropNetworking` to fail 🤔 I am not entirely sure though.

**Review by vdemeester — COMMENTED** [review]
@priyawadhwa @dlorenc any reason for what is happening in `dropNetworking` to be in the entrypoint and not outside (on the spec level — enforce by the controller through an annotation or something) ? Does using this feature requires that the user that runs the container (`USER` in Dockerfile, …) to be root, or something else ?

File: test/hermetic_taskrun_test.go (hunk 9 above)
Comment (chmouel) [reply test/hermetic_taskrun_test.go:35]: Not just OpenShift but i guess dropNetworking would be a privileged operation in general.

**Comment by chmouel** [comment]
If we open the door to netns manipulation from entrypoint or via api perhaps we can get userns too 😍

**Comment by dlorenc** [comment]
> @priyawadhwa @dlorenc any reason for what is happening in `dropNetworking` to be in the entrypoint and not outside (on the spec level — enforce by the controller through an annotation or something) ?

Not sure I understand this part - there's some discussion of other ways to do this in the TEP. The short answer is, there aren't any, or weren't when I last tried: https://github.com/tektoncd/community/blob/main/teps/0025-hermekton.md

> Does using this feature requires that the user that runs the container (`USER` in Dockerfile, …) to be root, or something else ?

There's some discussion on what is required here: https://github.com/tektoncd/pipeline/commit/b90a0efc4d7cfb4631a7ebb74de5f88b4df9dc2d#diff-e991ea79e58bc08e494f6f5f3bfbfeaa7d55b1bfbb68f90641500c1a7c890d21

In short - it's not exactly clear but it does work in most kernels. I think we can make it work without root ( if it doesn't already ) using some combination of user namespaces and setuid/dropping privleges on the binary itself.

**Comment by dlorenc** [comment]
/lgtm

nice!!!

**Comment by sbwsg** [comment]
The PR isn't merging and this might be either because we typically squash commits down to 1 prior to merge or it might mean a rebase is needed.

Suggest doing both of these and force pushing once more.  Hopefully once that's done tide should be content to merge.

**Comment by priyawadhwa** [comment]
Thanks for the tip @sbwsg, I've rebased & squashed!

**Comment by sbwsg** [comment]
re-adding the lgtm  🤞 

/lgtm

**Comment by priyawadhwa** [comment]
Yay! Thanks everyone!
